libs/libvtrutil/src/vtr_ragged_matrix.h

```diff
@@ -0,0 +1,223 @@
+    template<class Callback>
+    FlatRaggedMatrix(size_t nrows, Callback& row_length_callback, T default_value = T())
+        : FlatRaggedMatrix(RowLengthIterator<Callback>(0, row_length_callback),
+                           RowLengthIterator<Callback>(nrows, row_length_callback),
+                           default_value) {}
+
+    //Constructs matrix from a container of row lengths
+    template<class Container>
+    FlatRaggedMatrix(Container container, T default_value = T())
+        : FlatRaggedMatrix(std::begin(container), std::end(container), default_value) {}
+
+    //Constructs matrix from an iterator range, where the length of the range is
+    //the number of rows, and iterator values are the row lengths.
+    template<class Iter>
+    FlatRaggedMatrix(Iter row_size_first, Iter row_size_last, T default_value = T()) {
+        size_t nrows = std::distance(row_size_first, row_size_last);
+        first_elem_.resize(nrows + 1, -1); //+1 for sentinel
+
+        size_t nelem = 0;
+        size_t irow = 0;
+        for (Iter iter = row_size_first; iter != row_size_last; ++iter) {
+            first_elem_[irow] = nelem;
+
+            nelem += *iter;
+            ++irow;
+        }
+
+        //Sentinel
+        first_elem_[irow] = nelem;
+
+        data_.resize(nelem + 1, default_value); //+1 for sentinel
+    }
+
+  public: //Accessors
+    //Iterators to *all* elements
+    auto begin() {
+        return data_.begin();
+    }
+
+    auto end() {
+        if (empty()) {
+            return data_.end();
+        }
+        return data_.end() - 1;
+    }
+
+    auto begin() const {
+        return data_.begin();
+    }
+
+    auto end() const {
+        if (empty()) {
+            return data_.end();
+        }
+        return data_.end() - 1;
+    }
+
+    size_t size() const {
+        if (data_.empty()) {
+            return 0;
+        }
+        return data_.size() - 1; //-1 for sentinel
+    }
+
+    bool empty() const {
+        return size() == 0;
+    }
+
+    //Indexing operators for the first dimension
+    ProxyRow<T> operator[](Index0 i) {
+        int idx = size_t(i);
+        return ProxyRow<T>(&data_[first_elem_[idx]],
+                           &data_[first_elem_[idx + 1]]);
+    }
+
+    ProxyRow<const T> operator[](Index0 i) const {
+        int idx = size_t(i);
+        return ProxyRow<const T>(&data_[first_elem_[idx]],
+                                 &data_[first_elem_[idx + 1]]);
+    }
+
+    void clear() {
+        data_.clear();
+        first_elem_.clear();
+    }
+
+    void swap(FlatRaggedMatrix<T, Index0, Index1>& other) {
+        std::swap(data_, other.data_);
+        std::swap(first_elem_, other.first_elem_);
+    }
+
+    friend void swap(FlatRaggedMatrix<T, Index0, Index1>& lhs, FlatRaggedMatrix<T, Index0, Index1>& rhs) {
+        lhs.swap(rhs);
+    }
+
+  public: //Types
+    //Proxy class used to represent a 'row' in the matrix
+    template<typename U>
+    class ProxyRow {
```

Comment:
Can you just use array_view here, rather than defining a new thing?

Reply:
Yep, makes sense. I've updated it.